docs/architecture/adr-056-proving-amnesia-attack.md

```diff
@@ -0,0 +1,117 @@
+
+![](../imgs/tm-amnesia-attack.png)
+
+1. C1 and F send PREVOTE messages
+2. C1 sends PRECOMMIT for round 1
+3. A new round is started, C2 and F send PREVOTE messages
+4. C2 and F then send PRECOMMIT messages
+5. F breaks the lock and goes back and sends PRECOMMIT messages in round 1
+
+
+This creates a fork on the main chain.  Back to the past, another form of flip flopping, creates a light fork (capable of fooling those not involved in consensus), in a similar way, with F taking the precommits from C1 and forging a commit from them.
+
+## Decision
+
+As the distinction between these two attacks (amnesia and back to the past) can only be distinguished by confirming with all validators (to see if it is a full fork or a light fork), for the purpose of simplicity, these attacks will be treated as the same.
+
+Currently, the evidence reactor is used to simply broadcast and store evidence. Instead of perhaps creating a new reactor for the specific task of verifying these attacks, the current evidence reactor will be extended.
+
+The process begins with a light client receiving conflicting headers (in the future this could also be a full node during fast sync), which it sends to a full node to analyse. As part of [evidence handling](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-047-handling-evidence-from-light-client.md), this could be deduced into potential amnesia evidence
+
+```
+type PotentialAmnesiaEvidence struct {
+	V1 []*types.Vote
+	V2 []*types.Vote
+
+	timestamp time.Time
+}
+```
+
+*REMARK: We could also use `VoteSet`*
+
+*NOTE: this is almost identical to Duplicate Vote Evidence and although one is for different rounds and the other is for the same round, there is a case that these two could be combined. In addition, I believe batch processing to be more efficient than individual processing and would think all evidence that constitutes multiple attackers should be grouped together.*
+
+The evidence should contain the precommit votes for the intersection of validators that voted for both rounds. The votes should be all valid and the height and time that the infringement was made should be within:
+
+`Trusting period - Amnesia trial period - Gossip safety margin`
+
+With reference to the honest nodes, C1 and C2, in the schematic, C2 will not PRECOMMIT an earlier round, but it is likely, if a node in C1 were to recieve +2/3 PREVOTE's or PRECOMMIT's for a higher round, that it would remove the lock and PREVOTE and PRECOMMIT for the later round. Therefore, unfortunately it is not a case of simply punishing all nodes that have double voted in the `PotentialAmnesiaEvidence`.
+
+Instead we use the Proof of Lock Change (PoLC) referred to in the [consensus spec](https://github.com/tendermint/spec/blob/master/spec/consensus/consensus.md#terms). When an honest node votes again for a later round, in very rare cases, it will generate the PoLC and store it in the evidence reactor for a time equal to the `Trusting Period`
+
```

Comment:
As far as I understand, storing the PoLC is only necessary when in round 1 a node voted for block A and in round 2 it votes for block B (A!=B). In case it re-votes for the same block, it does not need to store anything.